inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/PdfAnnoRenderer.java

```diff
@@ -108,35 +110,62 @@ public static PdfAnnoModel render(AnnotatorState aState, VDocument aVDoc, String
-                cleanAnnotatedText.substring(start - index, end - index).replaceAll("\\s", ""),
-                first.getPosition(), last.getPosition());
+        int docTextLen = aDocumentText.length();
+        int vSpanBegin = aDocumentText.length();
+        int vSpanEnd = -1;
+        // search for begin of the first range and end of the last range and use those
+        for (VRange range : aVspan.getRanges())
+        {
+            if (range.getBegin() < vSpanBegin) vSpanBegin = range.getBegin();
+            if (range.getEnd() > vSpanEnd) vSpanEnd = range.getEnd();
         }
+
+        // use an offset before and after string to find a unique text snippet for an annotation
+        // begin with 0 and increase by 1 until a unique text is found
+        // offset is limited to length of a quarter of the text
+        for (int offset = 0; offset < docTextLen / 4 + 1; offset++)
+        {
+            // subtract offset from vSpanBegin and add offset to vSpanEnd and stay in bounds
+            int offsetBegin = vSpanBegin <= offset ? 0 : vSpanBegin - offset;
+            int offsetEnd = vSpanEnd < docTextLen - offset ? vSpanEnd + offset : docTextLen - 1;
+
+            // get annotated text from aDocumentText with offset before and after it
+            // also remove all whitespaces because they do not exist in PDFExtract text
+            String annotatedText = aDocumentText.substring(offsetBegin, offsetEnd).replaceAll("\\s", "");
+            String offsetBeforeText = aDocumentText.substring(offsetBegin, vSpanBegin).replaceAll("\\s", "");
+            String offsetAfterText = aDocumentText.substring(vSpanEnd, offsetEnd).replaceAll("\\s", "");
+
+            List<Integer> indices =
+                getAllIndices(aPdfExtractFile.getStringContent(), annotatedText);
+            if (indices.size() == 0) {
+                // if there are no matches end search as adding more characters won't help anymore
+                break;
+            } else if (indices.size() == 1) {
+                int index = indices.get(0);
+                // get begin and end position of the original annotationText within PDFExtract text
+                int annotationBegin = index + offsetBeforeText.length();
+                int annotationEnd = index + annotatedText.length() - offsetAfterText.length() - 1;
+                // get according PDFExtract file lines for begin and end of annotation
+                PdfExtractLine firstLine = aPdfExtractFile.getStringPdfExtractLine(annotationBegin);
+                PdfExtractLine lastLine = aPdfExtractFile.getStringPdfExtractLine(annotationEnd);
+                return new Span(aVspan.getVid().getId(), firstLine.getPage(), aColor,
+                    annotatedText.substring(annotationBegin - index, annotationEnd - index + 1),
+                    firstLine.getPosition(), lastLine.getPosition());
+            }
+        }
+        LOG.error("Could not map existing annotation with id " + aVspan.getVid().toString());
```

Comment:
I kind of assume that the case indices.size() > 1 means that there were multiple places where the annotation could match. It would be good to give different messages depending on whether the annotation could not be rendered because there were no matches and because there were multiple matches. The message should be displayed in the UI, not only in the error log - otherwise the user has no idea what is happening.

Reply:
Yes that's right.
If I'm not mistaken we don't need an error message for multiple matches.
If there are multiple matches we will find a unique match at latest when we search for annotation with whole document as context (with changes discussed earlier about offset limit to document text length).

Reply:
Ah... I think *now* I understand why earlier you increased this one value by 5 in every step. That wasn't a search offset but a context window size, wasn't it? I suppose increasing the window size by an increment > 1 could indeed make sense.

Reply:
Make it a constant with a speaking name like `WINDOW_SIZE_INCREMENT` - sure, 5 sounds good. Be sure to handle cases where the match is near the document boundaries and where you actually may not find the full number of chars before/after (or both).

Reply:
Ok, I will also rename `offset` to e.g. `windowSize`.
I already take care of boundaries when setting `offsetBegin` and `offsetEnd`. :)